trunner/test/harnesses/test_micropython.py

```diff
@@ -0,0 +1,503 @@
+        program_output = f"Running test: {test_path}\r\nSome result{PROMPT}"
+        exp_output = f"cat {test_path}.exp\r\nDifferent result{PROMPT}"
+
+        result = get_harness_output(MicropythonStandardHarness().harness,
+                                    TestStandardHarness.fake_program,
+                                    (test_path, program_output, exp_output))
+
+        exp_result = TestResult(test_path, TestResult.FAIL,
+                                "Incorrect result!\n\nExpected result:\nDifferent result\nTest result:\nSome result\n")
+
+        check_result(result, exp_result)
+
+    def test_special_test_pass(self):
+        # This path is defined in harness as special test
+        test_path = "/usr/test/micropython/test/special_test.py"
+
+        text = "aaa\r\n123\r\ngray\r\ngrey"
+        regex = "a\\{3\\}\r\n\\\\d\\+\r\ngr\\(a\\|e\\)y\r\ngr\\(a\\|e\\)y"
+
+        program_output = f"Running test: {test_path}\r\n{text}{PROMPT}"
+        exp_output = f"cat {test_path}.exp\r\n{regex}{PROMPT}"
+
+        result = get_harness_output(MicropythonStandardHarness().harness,
+                                    TestStandardHarness.fake_program,
+                                    (test_path, program_output, exp_output))
+
+        check_result(result, TestResult(test_path, TestResult.PASS))
+
+    def test_special_test_fail(self):
+        # This path is defined in harness as special test
+        test_path = "/usr/test/micropython/test/special_test.py"
+
+        text = "abc"
+        regex = "\\\\d\\+"
+
+        program_output = f"Running test: {test_path}\r\n{text}{PROMPT}"
+        exp_output = f"cat {test_path}.exp\r\n{regex}{PROMPT}"
+
+        result = get_harness_output(MicropythonStandardHarness().harness,
+                                    TestStandardHarness.fake_program,
+                                    (test_path, program_output, exp_output))
+
+        exp_result = TestResult(test_path, TestResult.FAIL,
+                                f"Incorrect result!\n\nExpected result:\n{regex}\nTest result:\n{text}\n")
+        check_result(result, exp_result)
+
+
+class TestReplHarness:
+
+    # Program for communicating with harness
+    def fake_program(socket, args):
+        test_path = args[0]
+        program_commands = args[1]
+        cmd_outputs = args[2]
+        exp_output = args[3]
+
+        if test_path is None:
+            return
+
+        # echo with test path
+        send_through_socket(socket, test_path + PROMPT)
+
+        # receiving cat with path to test commands
+        recv_exp_through_socket(socket, f"cat {test_path}\n")
+
+        if program_commands is None:
+            return
+
+        commands = []
+        for cmd in program_commands:
+            element = cmd
+
+            if not cmd == program_commands[-1]:
+                element += '\r\n'
+
+            commands += element
+
+        commands += PROMPT
+        commands = "".join(commands)
```

Comment:
```suggestion
        commands = "\r\n".join(program_commands)
        commands += PROMPT
        # or
        commands = f"{'\r\n'.join(program_commands)}{PROMPT}"
```
There is no need to build the entire string from array